src/Rematch.jl

```diff
@@ -112,11 +123,32 @@ function handle_destruct(value::Symbol, pattern, bound::Set{Symbol}, asserts::Ve
             end
         end
     elseif @capture(pattern, T_(subpatterns__))
-        @assert occursin(r"^[A-Z]", string(T)) "Pattern $pattern looks like a struct pattern but $T is probably not a struct type."
         # struct
-        push!(asserts, quote
-            assert_num_fields($(esc(T)), $(length(subpatterns)))
-        end)
+        @assert occursin(r"^[A-Z]", string(T)) "Pattern $pattern looks like a struct pattern but $T is probably not a struct type."
+        len = length(subpatterns)
+        named_fields = [pat.args[1] for pat in subpatterns if (pat isa Expr) && pat.head == :kw]
+        named_count = length(named_fields)
+        @assert named_count == length(unique(named_fields)) "Pattern $pattern has duplicate named arguments: $(named_fields)"
+        @assert named_count == 0 || len == named_count "Pattern $pattern mixes named and positional arguments"
+        if named_count == 0
+            # Uses positional variables
+            expected_fieldcount = gensym("$(T)_expected_fieldcount")
+            actual_fieldcount = gensym("$(T)_actual_fieldcount")
```

Comment:
Also, I actually don't think you need to call `gensym` here, as the macro system will do that for you automatically:
https://docs.julialang.org/en/v1/manual/metaprogramming/#Hygiene-1

That said, i suppose it doesn't _hurt_, but it maybe distracts from the important thing, which is that they're unique across _patterns_ -- ie the `$(T)` part of it.